ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/backwardsync/BackwardsSyncAlgorithm.java

```diff
@@ -0,0 +1,194 @@
+    return new ProcessKnownAncestorsStep(context, context.getBackwardChain()).executeAsync();
+  }
+
+  @VisibleForTesting
+  public CompletableFuture<Void> executeSyncStep(final Hash hash) {
+    return new SyncStepStep(context, context.getBackwardChain()).executeAsync(hash);
+  }
+
+  @VisibleForTesting
+  protected CompletableFuture<Void> executeBackwardAsync(final BlockHeader firstHeader) {
+    return new BackwardSyncStep(context, context.getBackwardChain()).executeAsync(firstHeader);
+  }
+
+  @VisibleForTesting
+  protected CompletableFuture<Void> executeForwardAsync() {
+    return new ForwardSyncStep(context, context.getBackwardChain()).executeAsync();
+  }
+
+  @VisibleForTesting
+  protected CompletableFuture<Void> waitForReady() {
+    final CountDownLatch latch = new CountDownLatch(1);
+    final long idTTD =
+        context.getSyncState().subscribeTTDReached(reached -> countDownIfReady(latch));
+    final long idIS =
+        context.getSyncState().subscribeCompletionReached(() -> countDownIfReady(latch));
+    return CompletableFuture.runAsync(() -> checkReadiness(latch, idTTD, idIS));
+  }
+
+  private void checkReadiness(final CountDownLatch latch, final long idTTD, final long idIS) {
+    try {
+      if (!context.isReady()) {
+        LOG.info("Waiting for preconditions...");
+        final boolean await = latch.await(2, TimeUnit.MINUTES);
+        if (await) {
+          LOG.info("Preconditions meet...");
+        }
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new BackwardSyncException("Wait for TTD preconditions interrupted");
+    } finally {
+      context.getSyncState().unsubscribeTTDReached(idTTD);
+      context.getSyncState().unsubscribeInitialConditionReached(idIS);
+    }
+  }
+
+  private void countDownIfReady(final CountDownLatch latch) {
+    if (context.isReady()) {
+      latch.countDown();
+    }
+  }
+
+  @VisibleForTesting
+  protected void runFinalizedSuccessionRule(
+      final MutableBlockchain blockchain, final Optional<Hash> maybeFinalized) {
+    if (maybeFinalized.isEmpty()) {
+      LOG.debug("Nothing to validate yet, consensus layer did not provide a new finalized block");
+      return;
+    }
+    final Hash newFinalized = maybeFinalized.get();
+    if (!blockchain.contains(newFinalized)) {
+      LOG.debug("New finalized block {} is not imported yet", newFinalized);
+      return;
+    }
+
+    final Optional<Hash> maybeOldFinalized = blockchain.getFinalized();
+    if (maybeOldFinalized.isPresent()) {
+      final Hash oldFinalized = maybeOldFinalized.get();
+      if (newFinalized.equals(oldFinalized)) {
+        LOG.debug("We already have this block as finalized");
+        return;
+      }
+      BlockHeader newFinalizedHeader =
+          blockchain
+              .getBlockHeader(newFinalized)
+              .orElseThrow(() -> new BackwardSyncException("Inconsistent Blockchain database...."));
```

Comment:
would be nice to have block header details in the exception message.

The same apply below